src/app/AttributePersistenceProvider.h

```diff
@@ -0,0 +1,94 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/ConcreteAttributePath.h>
+#include <app/util/attribute-metadata.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * Interface for persisting attribute values.
+ */
+
+class AttributePersistenceProvider
+{
+public:
+    virtual ~AttributePersistenceProvider() = default;
+    AttributePersistenceProvider()          = default;
+
+    /**
+     * Write an attribute value from the attribute store (i.e. not a struct or
+     * list) to non-volatile memory.
+     *
+     * @param [in] aPath the attribute path for the data being written.
+     * @param [in] aMetadata the attribute metadata, as a convenience.
+     * @param [in] aValue the data to write.  Integers and floats are
+     *             represented in native endianness.  Strings are represented
+     *             as Pascal-style strings, as in ZCL, with a length prefix
+     *             whose size depends on the actual string type.  The length is
+     *             stored as little-endian.
+     *
+     *             Integer and float values have a size that matches the `size`
+     *             member of aMetadata.
+     *
+     *             String values have a size that corresponds to the actual size
+     *             of the data in the string (including the length prefix),
+     *             which is no larger than the `size` member of aMetadata.
+     */
+    virtual CHIP_ERROR WriteValue(const ConcreteAttributePath & aPath, const EmberAfAttributeMetadata * aMetadata,
```

Comment:
I m curious about how the several chip manufacturers will implement this in their sdk
also how can we reroute this to external flash . Are they  aware of this implementation ? This will help them i guess

Reply:
As long as the storage delegate handed in here routes to flash this will work automatically, right?

Reply:
This answering my question and i take note for later.
by external flash i mean having 2 distinct locations of NVM storage : 
 1. Data from "a" regular frontend cluster in MCU's flash
 2. Data from an "OTA like" cluster on external flash (e.g accessible via SPI)  (non the MCU built-in flash)